### PR TITLE
Apt work

### PR DESCRIPTION
**Pull request by feliwir**
- Parse PlaceObjectActions (action that a triggered under certain conditions)
- Add new instructions
- Put builtin ActionScript functions/objects into a seperate class

### CODE REVIEW

File: src/OpenSage.Game/Data/Apt/Characters/Button.cs [hunk 1]
```diff
@@ -44,7 +44,8 @@ public struct ButtonRecord
         public static ButtonRecord Parse(BinaryReader reader)
         {
             return new ButtonRecord(
-                reader.ReadUInt32AsEnumFlags<ButtonRecordFlags>(),
+                reader.ReadByteAsEnumFlags<ButtonRecordFlags>(),
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Characters/Button.cs]: In the binary file, is the enum actually stored in a uint32? If so, I think this was better before this change.

File: src/OpenSage.Game/Data/Apt/Characters/Button.cs [hunk 2]
```diff
@@ -123,7 +124,7 @@ public static Button Parse(BinaryReader reader)
         {
             var button = new Button();
 
-            button.IsMenu = reader.ReadBooleanUInt32Checked();
+            button.IsMenu = Convert.ToBoolean(reader.ReadUInt32());
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Apt/Characters/Button.cs]: What's the reason for this change?

File: src/OpenSage.Game/Data/Utilities/Extensions/BinaryReaderExtensions.cs [hunk 3]
```diff
@@ -396,6 +404,28 @@ public static List<T> ReadListAtOffset<T>(this BinaryReader reader, Func<T> crea
             return result;
         }
 
+        public static List<T> ReadFixedSizeListAtOffset<T>(this BinaryReader reader, Func<T> creator, uint size) where T : class
```
Comment (tgjones) [inline src/OpenSage.Game/Data/Utilities/Extensions/BinaryReaderExtensions.cs]: What's the reason for this, in addition to `ReadFixedSizeArrayAtOffset`?

File: src/OpenSage.Game/Graphics/SpriteBatch.cs [hunk 4]
```diff
@@ -73,7 +73,7 @@ public SpriteBatch(ContentManager contentManager)
 
         private ref SpriteBatchItem CreateBatchItem()
         {
-            if (_currentBatchIndex > _batchItems.Length)
+            if (_currentBatchIndex >= _batchItems.Length)
```
Comment (tgjones) [inline src/OpenSage.Game/Graphics/SpriteBatch.cs]: Thanks :)

File: src/OpenSage.Game/Gui/Apt/ActionScript/Function.cs [hunk 5]
```diff
@@ -1,13 +1,36 @@
...
+﻿using System;
+using System.Collections.Generic;
 using OpenSage.Gui.Apt.ActionScript.Opcodes;
 
 namespace OpenSage.Gui.Apt.ActionScript
 {
+    [Flags]
+    public enum FunctionPreloadFlags
+    {
+        PreloadExtern = 0x010000,   //this seems to be added by EA
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/Function.cs]: It's not critical, but I usually like enum values to be in ascending numeric order, unless there's a good reason not to.

File: src/OpenSage.Game/Gui/Apt/ActionScript/Opcodes/Object.cs [hunk 6]
```diff
@@ -100,18 +117,44 @@ public override void Execute(ActionContext context)
...
     public sealed class GetProperty : InstructionBase
     {
         public override InstructionType Type => InstructionType.GetProperty;
 
         public override void Execute(ActionContext context)
         {
-            throw new NotImplementedException();
+            var property = context.Stack.Pop();
+            var target = context.GetTarget(context.Stack.Pop().ToString());
+            var type = (PropertyType) property.ToInteger();
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/Opcodes/Object.cs]: If this is a common thing to do, would it be worth having a `property.ToEnum<TEnum>()` method?

File: src/OpenSage.Game/Data/Apt/Characters/Button.cs (hunk 1 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Apt/Characters/Button.cs]: Couldn't do it like before, because i encountered some other bits in the uint24. I don't know what those mean yet, but it makes the Enum verification fail

File: src/OpenSage.Game/Data/Apt/Characters/Button.cs (hunk 2 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Apt/Characters/Button.cs]: I have no idea, why i did that :)

File: src/OpenSage.Game/Data/Utilities/Extensions/BinaryReaderExtensions.cs (hunk 3 above)
Comment (feliwir) [reply src/OpenSage.Game/Data/Utilities/Extensions/BinaryReaderExtensions.cs]: I needed it for the new DefineFunction2. I guess i should decide wether i use Array or List at some point

File: src/OpenSage.Game/Gui/Apt/ActionScript/Function.cs (hunk 5 above)
Comment (feliwir) [reply src/OpenSage.Game/Gui/Apt/ActionScript/Function.cs]: I did that because that's the order of bits. Otherwise i always get confused

File: src/OpenSage.Game/Gui/Apt/ActionScript/Opcodes/Object.cs (hunk 6 above)
Comment (feliwir) [reply src/OpenSage.Game/Gui/Apt/ActionScript/Opcodes/Object.cs]: property is of type Value, where i could add that. But the Value type has lots of room for improvements anyways ;D

**Comment by tgjones** [comment]
Great work, thank you :)
